src/R2_alg.jl

```diff
@@ -167,53 +253,94 @@ function R2(
+    verbose = options.verbose,
+    max_iter = options.maxIter,
+    max_time = options.maxTime,
+    σmin = options.σmin,
+    η1 = options.η1,
+    η2 = options.η2,
+    ν = options.ν,
+    γ = options.γ,
+  )
   outdict = Dict(
-    :Fhist => solver.Fobj_hist[1:k],
-    :Hhist => solver.Hobj_hist[1:k],
-    :Chist => solver.Complex_hist[1:k],
+    :Fhist => stats.solver_specific[:Fhist],
+    :Hhist => stats.solver_specific[:Hhist],
+    :Chist => stats.solver_specific[:SubsolverCounter],
     :NonSmooth => h,
-    :status => status,
-    :fk => fk,
-    :hk => hk,
-    :ξ => ξ,
-    :elapsed_time => elapsed_time,
+    :status => stats.status,
+    :fk => stats.solver_specific[:smooth_obj],
+    :hk => stats.solver_specific[:nonsmooth_obj],
+    :ξ => stats.solver_specific[:xi],
+    :elapsed_time => stats.elapsed_time,
   )
-  return solver.xk, k, outdict
+  return stats.solution,stats.iter,outdict
 end
 
-function R2!(
-  solver::R2Solver{R},
-  f::F,
-  ∇f!::G,
-  h::H,
-  options::ROSolverOptions{R},
-  x0::AbstractVector{R};
-  selected::AbstractVector{<:Integer} = 1:length(x0),
-) where {F <: Function, G <: Function, H, R <: Real}
-  start_time = time()
-  elapsed_time = 0.0
-  ϵ = options.ϵa
-  ϵr = options.ϵr
-  neg_tol = options.neg_tol
-  verbose = options.verbose
-  maxIter = options.maxIter
-  maxTime = options.maxTime
-  σmin = options.σmin
-  η1 = options.η1
-  η2 = options.η2
-  ν = options.ν
-  γ = options.γ
-
-  # retrieve workspace
-  xk = solver.xk
-  xk .= x0
+
+function R2(reg_nlp::AbstractRegularizedNLPModel; kwargs...)
+  solver = R2Solver(reg_nlp)
+  stats = GenericExecutionStats(reg_nlp.model)
+  cb = (nlp, solver, stats) -> begin
+    solver.Fobj_hist[stats.iter+1] = stats.solver_specific[:smooth_obj]
+    solver.Hobj_hist[stats.iter+1] = stats.solver_specific[:nonsmooth_obj]
+    solver.Complex_hist[stats.iter+1] += 1
+  end
+  solve!(
+    solver,
+    reg_nlp,
+    stats;
+    callback = cb,
+    kwargs...
+  )
+  set_solver_specific!(stats, :Fhist, solver.Fobj_hist[1:stats.iter+1])
+  set_solver_specific!(stats, :Hhist, solver.Hobj_hist[1:stats.iter+1])
+  set_solver_specific!(stats, :SubsolverCounter, solver.Complex_hist[1:stats.iter+1])
+  return stats
+end
+
+function SolverCore.solve!(
+  solver::R2Solver{T}, 
+  reg_nlp::AbstractRegularizedNLPModel{T, V}, 
+  stats::GenericExecutionStats{T, V};
+  callback = (args...) -> nothing,
+  x::V = reg_nlp.model.meta.x0,
+  atol::T = √eps(T),
+  rtol::T = √eps(T),
+  neg_tol::T = eps(T)^(1 / 4),
+  verbose::Int = 0,
+  max_iter::Int = 10000,
+  max_time::Float64 = 30.0,
+  max_eval::Int = -1,
+  σmin::T = eps(T),
+  η1::T = √√eps(T),
+  η2::T = T(0.9),
+  ν::T = eps(T)^(1 / 5),
+  γ::T = T(3),
+  ) where {T, V}
+
+  reset!(stats)
+
+  # Retrieve workspace
+  selected = reg_nlp.selected
+  h = reg_nlp.h
+  nlp = reg_nlp.model
+  
+  # Make sure ψ is well shifted (on nlp.meta.x0 in R2Solver Constructor but might be different)
+  if solver.xk != x
```

Comment:
If you want to check that the vectors are the same (in the sense that the components are the same), use
```julia
if !all(solver.xk .== x)
```